File: packages/infra/browser-container/src/screencast-manager.ts
/**
 * Screencast Manager
 * Handles CDP-based browser screencasting with bandwidth optimization
 */

import { CDPSession, Page } from 'playwright-core';
import { WebSocket } from 'ws';
import { createHash } from 'crypto';
import type {
    ScreencastOptions,
    ScreencastMetadata,
    FrameDetectionState,
    InputEvent,
    ScreencastFrame,
    ScreencastStats
} from './types';

export interface ScreencastSession {
    sessionId: string;
    cdpSession: CDPSession;
    page: Page;
    websocket: WebSocket;
    detectionState: FrameDetectionState;
    stats: ScreencastStats;
    isActive: boolean;
}

export class ScreencastManager {
    private activeSessions = new Map<string, ScreencastSession>();
    private readonly defaultOptions: Required<ScreencastOptions> = {
        quality: 80,
        everyNthFrame: 10,
        detectIdle: true,
        idleThreshold: 0.01,
        maxWidth: 1024,
        maxHeight: 768
    };

    /**
     * Start screencast for a session
     */
    async startScreencast(
        sessionId: string,
        page: Page,
        websocket: WebSocket,
        options: ScreencastOptions = {}
    ): Promise<void> {
        if (this.activeSessions.has(sessionId)) {
            throw new Error(`Screencast already active for session ${sessionId}`);
        }

        const finalOptions = { ...this.defaultOptions, ...options };

        // Create CDP session for this page
        const cdpSession = await page.context().newCDPSession(page);

        const detectionState: FrameDetectionState = {
            lastFrameHash: null,
            idleFrameCount: 0,
            lastForcedTime: Date.now(),
            options: finalOptions
        };

        const stats: ScreencastStats = {
            framesSent: 0,
            framesSkipped: 0,
            bytesTransmitted: 0,
            averageFrameSize: 0,
            actualFps: 0,
            skipPercentage: 0
        };

        const screencastSession: ScreencastSession = {
            sessionId,
            cdpSession,
            page,
            websocket,
            detectionState,
            stats,
            isActive: true
        };

        this.activeSessions.set(sessionId, screencastSession);

        try {
            // Start CDP screencast
            await cdpSession.send('Page.startScreencast', {
                format: 'jpeg',
                quality: finalOptions.quality,
                everyNthFrame: finalOptions.everyNthFrame,
                maxWidth: finalOptions.maxWidth,
                maxHeight: finalOptions.maxHeight
            });

            // Listen for screencast frames
            cdpSession.on('Page.screencastFrame', async (frameData) => {
                await this.handleScreencastFrame(sessionId, frameData);
            });

            console.log(`Screencast started for session ${sessionId}`);

            // Send success message
            this.sendMessage(websocket, {
                type: 'SCREENCAST_STARTED',
                sessionId,
                options: finalOptions
            });

        } catch (error) {
            this.activeSessions.delete(sessionId);
            throw error;
        }
    }

    /**
     * Stop screencast for a session
     */
    async stopScreencast(sessionId: string): Promise<void> {
        const session = this.activeSessions.get(sessionId);
        if (!session) {
            return; // Already stopped
        }

        try {
            await session.cdpSession.send('Page.stopScreencast');
            session.isActive = false;

            // Remove listeners - just detach the CDP session
            await session.cdpSession.detach();

            console.log(`Screencast stopped for session ${sessionId}`, {
                stats: session.stats
            });

            // Send stop message
            this.sendMessage(session.websocket, {
                type: 'SCREENCAST_STOPPED',
                sessionId,
                stats: session.stats
            });

        } catch (error) {
            console.error(`Error stopping screencast for session ${sessionId}:`, error);
        } finally {
            this.activeSessions.delete(sessionId);
        }
    }

    /**
     * Handle user input events
     */
    async handleInput(sessionId: string, inputEvent: InputEvent): Promise<void> {
        const session = this.activeSessions.get(sessionId);
        if (!session) {
            console.warn(`No active screencast session for input: ${sessionId}`);
            return;
        }

        try {
            const { cdpSession } = session;

            switch (inputEvent.type) {
                case 'mousePressed':
                case 'mouseReleased':
                case 'mouseMoved':
                    await cdpSession.send('Input.dispatchMouseEvent', {
                        type: inputEvent.type,
                        x: inputEvent.x || 0,
                        y: inputEvent.y || 0,
                        button: (inputEvent.button as any) || 'left',
                        clickCount: inputEvent.clickCount || 1,
                        modifiers: inputEvent.modifiers || 0
                    });
                    break;

                case 'keyDown':
                case 'keyUp':
                    await cdpSession.send('Input.dispatchKeyEvent', {
                        type: inputEvent.type,
                        key: inputEvent.key,
                        code: inputEvent.code,
                        text: inputEvent.text,
                        modifiers: inputEvent.modifiers || 0
                    });
                    break;

                case 'char':
                    await cdpSession.send('Input.dispatchKeyEvent', {
                        type: inputEvent.type,
                        text: inputEvent.text,
                        modifiers: inputEvent.modifiers || 0
                    });
                    break;

                case 'mouseWheel':
                    await cdpSession.send('Input.dispatchMouseEvent', {
                        type: 'mouseWheel',
                        x: inputEvent.x || 0,
                        y: inputEvent.y || 0,
                        deltaX: inputEvent.deltaX || 0,
                        deltaY: inputEvent.deltaY || 0
                    });
                    break;

                default:
                    console.warn(`Unknown input event type: ${(inputEvent as any).type}`);
            }

            // Reset idle detection state on user input to force next frame
            session.detectionState.lastFrameHash = null;

        } catch (error) {
            console.error(`Error handling input for session ${sessionId}:`, error);
        }
    }

    /**
     * Handle incoming screencast frame from CDP
     */
    private async handleScreencastFrame(
        sessionId: string,
        frameData: any
    ): Promise<void> {
        const session = this.activeSessions.get(sessionId);
        if (!session || !session.isActive) {
            return;
        }

        try {
            const { detectionState, stats } = session;

            // Check if we should send this frame
            const shouldSend = await this.shouldSendFrame(frameData.data, detectionState);

            if (shouldSend) {
                // Send frame to client
                const frame: ScreencastFrame = {
                    data: frameData.data,
                    metadata: frameData.metadata as ScreencastMetadata,
                    sessionId,
                    frameId: frameData.sessionId
                };

                this.sendMessage(session.websocket, {
                    type: 'SCREENCAST_FRAME',
                    ...frame
                });

                // Update stats
                stats.framesSent++;
                const frameSize = Buffer.byteLength(frameData.data, 'base64');
                stats.bytesTransmitted += frameSize;
                stats.averageFrameSize = stats.bytesTransmitted / stats.framesSent;
            } else {
                stats.framesSkipped++;
            }

            // Update skip percentage
            const totalFrames = stats.framesSent + stats.framesSkipped;
            stats.skipPercentage = totalFrames > 0 ? (stats.framesSkipped / totalFrames) * 100 : 0;

            // Always acknowledge frame to CDP
            await session.cdpSession.send('Page.screencastFrameAck', {
                sessionId: frameData.sessionId
            });

        } catch (error) {
            console.error(`Error handling screencast frame for session ${sessionId}:`, error);
        }
    }

    /**
     * Determine if frame should be sent based on change detection
     */
    private async shouldSendFrame(
        frameData: string,
        state: FrameDetectionState
    ): Promise<boolean> {
        if (!state.options.detectIdle) {
            return true; // Send all frames if idle detection is disabled
        }

        try {
            const currentHash = this.computeFrameHash(frameData);
            const now = Date.now();

            // Always send first frame
            if (state.lastFrameHash === null) {
                state.lastFrameHash = currentHash;
                return true;
            }

            // Check for visual changes
            const hasChanged = currentHash !== state.lastFrameHash;

            // Force frame every 5 seconds to keep connection alive
            const shouldForce = (now - state.lastForcedTime) > 5000;

            if (hasChanged || shouldForce) {
                state.lastFrameHash = currentHash;
                state.idleFrameCount = 0;
                if (shouldForce) {
                    state.lastForcedTime = now;
                }
                return true;
            }

            // Skip idle frame
            state.idleFrameCount++;
            return false;

        } catch (error) {
            console.error('Error in frame change detection, sending frame:', error);
            return true; // Fall back to sending frame on error
        }
    }

    /**
     * Compute hash of frame data for change detection
     */
    private computeFrameHash(frameData: string): string {
        // Create hash of base64 data
        return createHash('md5').update(frameData).digest('hex');
    }

    /**
     * Send message over WebSocket
     */
    private sendMessage(websocket: WebSocket, message: any): void {
        if (websocket.readyState === WebSocket.OPEN) {
            websocket.send(JSON.stringify(message));
        }
    }

    /**
     * Get stats for a session
     */
    getSessionStats(sessionId: string): ScreencastStats | null {
        const session = this.activeSessions.get(sessionId);
        return session ? { ...session.stats } : null;
    }

    /**
     * Get all active session IDs
     */
    getActiveSessions(): string[] {
        return Array.from(this.activeSessions.keys());
    }

    /**
     * Stop all active screencasts (cleanup)
     */
    async stopAllScreencasts(): Promise<void> {
        const stopPromises = Array.from(this.activeSessions.keys()).map(sessionId =>
            this.stopScreencast(sessionId)
        );
        await Promise.all(stopPromises);
    }

    /**
     * Check if session has active screencast
     */
    isScreencastActive(sessionId: string): boolean {
        const session = this.activeSessions.get(sessionId);
        return session ? session.isActive : false;
    }
} 

File: packages/infra/browser-container/screencast_manager.go
package main

import (
	"context"
	"crypto/md5"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"strings"
	"sync"
	"time"

	"github.com/gorilla/websocket"
)

// ScreencastOptions represents screencast configuration
type ScreencastOptions struct {
	Quality        int  `json:"quality,omitempty"`
	EveryNthFrame  int  `json:"everyNthFrame,omitempty"`
	DetectIdle     bool `json:"detectIdle,omitempty"`
	IdleThreshold  int  `json:"idleThreshold,omitempty"`
	MaxWidth       int  `json:"maxWidth,omitempty"`
	MaxHeight      int  `json:"maxHeight,omitempty"`
}

// ScreencastMetadata represents frame metadata
type ScreencastMetadata struct {
	OffsetTop      float64 `json:"offsetTop"`
	PageScaleFactor float64 `json:"pageScaleFactor"`
	DeviceWidth    int     `json:"deviceWidth"`
	DeviceHeight   int     `json:"deviceHeight"`
	ScrollOffsetX  float64 `json:"scrollOffsetX"`
	ScrollOffsetY  float64 `json:"scrollOffsetY"`
	Timestamp      int64   `json:"timestamp"`
}

// ScreencastFrame represents a screencast frame
type ScreencastFrame struct {
	Data      string             `json:"data"`
	Metadata  ScreencastMetadata `json:"metadata"`
	SessionID string             `json:"sessionId"`
	FrameID   int                `json:"frameId"`
}

// FrameDetectionState tracks frame detection for idle detection
type FrameDetectionState struct {
	LastFrameHash   string                `json:"lastFrameHash"`
	IdleFrameCount  int                   `json:"idleFrameCount"`
	LastForcedTime  int64                 `json:"lastForcedTime"`
	Options         ScreencastOptions     `json:"options"`
}

// ScreencastStats tracks performance statistics
type ScreencastStats struct {
	FramesSent       int     `json:"framesSent"`
	FramesSkipped    int     `json:"framesSkipped"`
	BytesTransmitted int64   `json:"bytesTransmitted"`
	AverageFrameSize int     `json:"averageFrameSize"`
	ActualFPS        float64 `json:"actualFps"`
	SkipPercentage   float64 `json:"skipPercentage"`
}

// ScreencastSession represents an active screencast session
type ScreencastSession struct {
	SessionID       string                   `json:"sessionId"`
	WebSocketConn   *websocket.Conn          `json:"-"`
	Context         context.Context          `json:"-"`
	Cancel          context.CancelFunc       `json:"-"`
	Options         ScreencastOptions        `json:"options"`
	Stats           ScreencastStats          `json:"stats"`
	DetectionState  FrameDetectionState      `json:"detectionState"`
	FrameCounter    int                      `json:"frameCounter"`
	StartTime       time.Time                `json:"startTime"`
	LastFrameTime   time.Time                `json:"lastFrameTime"`
	mutex           sync.RWMutex             `json:"-"`
}

// ScreencastManager manages screencast sessions
type ScreencastManager struct {
	sessions      map[string]*ScreencastSession
	sessionsMutex sync.RWMutex
}

// NewScreencastManager creates a new screencast manager
func NewScreencastManager() *ScreencastManager {
	return &ScreencastManager{
		sessions: make(map[string]*ScreencastSession),
	}
}

// StartScreencast starts screencasting for a session
func (sm *ScreencastManager) StartScreencast(sessionID string, wsConn *websocket.Conn, options *ScreencastOptions) error {
	sm.sessionsMutex.Lock()
	defer sm.sessionsMutex.Unlock()

	// Check if screencast is already active for this session
	if _, exists := sm.sessions[sessionID]; exists {
		return fmt.Errorf("screencast already active for session %s", sessionID)
	}

	// Set default options
	if options == nil {
		options = &ScreencastOptions{
			Quality:       80,
			EveryNthFrame: 1,
			DetectIdle:    true,
			IdleThreshold: 5,
			MaxWidth:      1920,
			MaxHeight:     1080,
		}
	}

	// Create session context
	ctx, cancel := context.WithCancel(context.Background())

	session := &ScreencastSession{
		SessionID:     sessionID,
		WebSocketConn: wsConn,
		Context:       ctx,
		Cancel:        cancel,
		Options:       *options,
		Stats:         ScreencastStats{},
		DetectionState: FrameDetectionState{
			Options: *options,
		},
		FrameCounter:  0,
		StartTime:     time.Now(),
		LastFrameTime: time.Now(),
	}

	sm.sessions[sessionID] = session

	// Start the screencast goroutine
	go sm.captureFrames(session)

	log.Printf("Screencast started for session: %s", sessionID)
	return nil
}

// StopScreencast stops screencasting for a session
func (sm *ScreencastManager) StopScreencast(sessionID string) error {
	sm.sessionsMutex.Lock()
	defer sm.sessionsMutex.Unlock()

	session, exists := sm.sessions[sessionID]
	if !exists {
		return fmt.Errorf("no active screencast for session %s", sessionID)
	}

	// Cancel the context to stop the capture goroutine
	session.Cancel()

	// Send final stats
	sm.sendScreencastStats(session)

	// Remove from sessions
	delete(sm.sessions, sessionID)

	log.Printf("Screencast stopped for session: %s", sessionID)
	return nil
}

// IsScreencastActive checks if screencast is active for a session
func (sm *ScreencastManager) IsScreencastActive(sessionID string) bool {
	sm.sessionsMutex.RLock()
	defer sm.sessionsMutex.RUnlock()
	_, exists := sm.sessions[sessionID]
	return exists
}

// GetScreencastStats returns stats for a session
func (sm *ScreencastManager) GetScreencastStats(sessionID string) (*ScreencastStats, error) {
	sm.sessionsMutex.RLock()
	defer sm.sessionsMutex.RUnlock()

	session, exists := sm.sessions[sessionID]
	if !exists {
		return nil, fmt.Errorf("no active screencast for session %s", sessionID)
	}

	session.mutex.RLock()
	defer session.mutex.RUnlock()
	stats := session.Stats
	return &stats, nil
}

// captureFrames is the main capture loop for a screencast session
func (sm *ScreencastManager) captureFrames(session *ScreencastSession) {
	defer func() {
		log.Printf("Capture loop ended for session: %s", session.SessionID)
	}()

	// Calculate frame interval (FPS = 30, so ~33ms between frames)
	frameInterval := time.Millisecond * 33

	ticker := time.NewTicker(frameInterval)
	defer ticker.Stop()

	for {
		select {
		case <-session.Context.Done():
			return
		case <-ticker.C:
			if err := sm.captureAndSendFrame(session); err != nil {
				log.Printf("Error capturing frame for session %s: %v", session.SessionID, err)
				// Continue on error - don't stop the entire screencast
			}
		}
	}
}

// captureAndSendFrame captures a single frame and sends it via WebSocket
func (sm *ScreencastManager) captureAndSendFrame(session *ScreencastSession) error {
	session.mutex.Lock()
	defer session.mutex.Unlock()

	// Check if we should skip this frame based on everyNthFrame
	if session.FrameCounter%session.Options.EveryNthFrame != 0 {
		session.FrameCounter++
		session.Stats.FramesSkipped++
		return nil
	}

	// Simulate frame capture (in real implementation, this would use chromedp)
	frameData := sm.simulateFrameCapture(session)

	// Check for idle detection if enabled
	if session.Options.DetectIdle && sm.isFrameIdle(session, frameData) {
		session.Stats.FramesSkipped++
		return nil
	}

	// Create frame metadata
	metadata := ScreencastMetadata{
		OffsetTop:       0,
		PageScaleFactor: 1.0,
		DeviceWidth:     session.Options.MaxWidth,
		DeviceHeight:    session.Options.MaxHeight,
		ScrollOffsetX:   0,
		ScrollOffsetY:   0,
		Timestamp:       time.Now().UnixMilli(),
	}

	// Create screencast frame
	frame := ScreencastFrame{
		Data:      frameData,
		Metadata:  metadata,
		SessionID: session.SessionID,
		FrameID:   session.FrameCounter,
	}

	// Send frame via WebSocket
	if err := sm.sendFrame(session, frame); err != nil {
		return err
	}

	// Update stats
	session.Stats.FramesSent++
	session.Stats.BytesTransmitted += int64(len(frameData))
	session.FrameCounter++
	session.LastFrameTime = time.Now()

	// Calculate average frame size
	if session.Stats.FramesSent > 0 {
		session.Stats.AverageFrameSize = int(session.Stats.BytesTransmitted / int64(session.Stats.FramesSent))
	}

	// Calculate actual FPS
	elapsed := time.Since(session.StartTime).Seconds()
	if elapsed > 0 {
		session.Stats.ActualFPS = float64(session.Stats.FramesSent) / elapsed
	}

	// Calculate skip percentage
	totalFrames := session.Stats.FramesSent + session.Stats.FramesSkipped
	if totalFrames > 0 {
		session.Stats.SkipPercentage = float64(session.Stats.FramesSkipped) / float64(totalFrames) * 100
	}

	return nil
}

// simulateFrameCapture simulates frame capture (placeholder for chromedp implementation)
func (sm *ScreencastManager) simulateFrameCapture(session *ScreencastSession) string {
	// In real implementation, this would use chromedp to capture a screenshot
	// For now, return a placeholder base64 encoded image
	placeholder := fmt.Sprintf("frame-%s-%d", session.SessionID, session.FrameCounter)
	return base64.StdEncoding.EncodeToString([]byte(placeholder))
}

// isFrameIdle checks if the current frame is idle (same as previous frame)
func (sm *ScreencastManager) isFrameIdle(session *ScreencastSession, frameData string) bool {
	// Calculate frame hash for comparison
	hasher := md5.New()
	hasher.Write([]byte(frameData))
	frameHash := fmt.Sprintf("%x", hasher.Sum(nil))

	// Compare with last frame
	if session.DetectionState.LastFrameHash == frameHash {
		session.DetectionState.IdleFrameCount++
		
		// Check if we've exceeded idle threshold
		if session.DetectionState.IdleFrameCount >= session.Options.IdleThreshold {
			// Check if we should force a frame (e.g., every 5 seconds)
			now := time.Now().UnixMilli()
			if now-session.DetectionState.LastForcedTime > 5000 {
				session.DetectionState.LastForcedTime = now
				session.DetectionState.IdleFrameCount = 0
				return false // Send this frame
			}
			return true // Skip this frame
		}
	} else {
		// Frame changed, reset idle count
		session.DetectionState.IdleFrameCount = 0
		session.DetectionState.LastFrameHash = frameHash
	}

	return false
}

// sendFrame sends a frame via WebSocket
func (sm *ScreencastManager) sendFrame(session *ScreencastSession, frame ScreencastFrame) error {
	if session.WebSocketConn == nil {
		return fmt.Errorf("no WebSocket connection for session %s", session.SessionID)
	}

	message := map[string]interface{}{
		"type":    "SCREENCAST_FRAME",
		"frame":   frame,
		"stats":   session.Stats,
	}

	return session.WebSocketConn.WriteJSON(message)
}

// sendScreencastStats sends final stats when screencast stops
func (sm *ScreencastManager) sendScreencastStats(session *ScreencastSession) {
	if session.WebSocketConn == nil {
		return
	}

	message := map[string]interface{}{
		"type":      "SCREENCAST_STATS",
		"sessionId": session.SessionID,
		"stats":     session.Stats,
		"duration":  time.Since(session.StartTime).Seconds(),
	}

	session.WebSocketConn.WriteJSON(message)
}

// HandleInput handles user input events for screencast sessions
func (sm *ScreencastManager) HandleInput(sessionID string, event *InputEvent) error {
	sm.sessionsMutex.RLock()
	session, exists := sm.sessions[sessionID]
	sm.sessionsMutex.RUnlock()

	if !exists {
		return fmt.Errorf("no active screencast for session %s", sessionID)
	}

	// Send input event notification
	if session.WebSocketConn != nil {
		message := map[string]interface{}{
			"type":      "INPUT_EVENT",
			"sessionId": sessionID,
			"event":     event,
		}
		session.WebSocketConn.WriteJSON(message)
	}

	return nil
}

// Methods referenced in main.go that need to be implemented
func (c *MultiSessionContainer) handleStartScreencast(w http.ResponseWriter, r *http.Request) {
	// Extract session ID from URL path
	sessionID := extractSessionIDFromPath(r.URL.Path, "/sessions/", "/start-screencast")
	if sessionID == "" {
		http.Error(w, "Invalid session ID", http.StatusBadRequest)
		return
	}

	// Parse options from request body
	var options ScreencastOptions
	if err := json.NewDecoder(r.Body).Decode(&options); err != nil {
		// Use default options if parsing fails
		options = ScreencastOptions{
			Quality:       80,
			EveryNthFrame: 1,
			DetectIdle:    true,
			IdleThreshold: 5,
			MaxWidth:      1920,
			MaxHeight:     1080,
		}
	}

	// Get the WebSocket connection for this session
	c.sessionWSMutex.RLock()
	wsConn, exists := c.sessionWS[sessionID]
	c.sessionWSMutex.RUnlock()

	if !exists {
		http.Error(w, "No WebSocket connection for session", http.StatusBadRequest)
		return
	}

	// Start screencast
	if err := c.screencastMgr.StartScreencast(sessionID, wsConn, &options); err != nil {
		http.Error(w, fmt.Sprintf("Failed to start screencast: %v", err), http.StatusInternalServerError)
		return
	}

	// Return success response with WebSocket URL
	response := map[string]interface{}{
		"success":       true,
		"sessionId":     sessionID,
		"screencastUrl": fmt.Sprintf("ws://localhost:%d/internal/ws?token=your-token", Port),
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(response)
}

func (c *MultiSessionContainer) handleStopScreencast(w http.ResponseWriter, r *http.Request) {
	// Extract session ID from URL path
	sessionID := extractSessionIDFromPath(r.URL.Path, "/sessions/", "/stop-screencast")
	if sessionID == "" {
		http.Error(w, "Invalid session ID", http.StatusBadRequest)
		return
	}

	// Stop screencast
	if err := c.screencastMgr.StopScreencast(sessionID); err != nil {
		http.Error(w, fmt.Sprintf("Failed to stop screencast: %v", err), http.StatusInternalServerError)
		return
	}

	response := map[string]interface{}{
		"success":   true,
		"sessionId": sessionID,
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(response)
}

func (c *MultiSessionContainer) handleStartScreencastMessage(conn *websocket.Conn, msg *InternalMessage) {
	options := ScreencastOptions{
		Quality:       80,
		EveryNthFrame: 1,
		DetectIdle:    true,
		IdleThreshold: 5,
		MaxWidth:      1920,
		MaxHeight:     1080,
	}

	if msg.Params != nil {
		// Parse options from params
		if quality, ok := msg.Params["quality"].(float64); ok {
			options.Quality = int(quality)
		}
		if everyNth, ok := msg.Params["everyNthFrame"].(float64); ok {
			options.EveryNthFrame = int(everyNth)
		}
	}

	if err := c.screencastMgr.StartScreencast(msg.SessionID, conn, &options); err != nil {
		conn.WriteJSON(map[string]interface{}{
			"type":      "SCREENCAST_ERROR",
			"sessionId": msg.SessionID,
			"error":     err.Error(),
		})
	}
}

func (c *MultiSessionContainer) handleStopScreencastMessage(msg *InternalMessage) {
	c.screencastMgr.StopScreencast(msg.SessionID)
}

func (c *MultiSessionContainer) handleInputEvent(msg *InternalMessage) {
	if msg.Event != nil {
		c.screencastMgr.HandleInput(msg.SessionID, msg.Event)
	}
}

// Helper function to extract session ID from URL path
func extractSessionIDFromPath(path, prefix, suffix string) string {
	if !strings.HasPrefix(path, prefix) || !strings.HasSuffix(path, suffix) {
		return ""
	}
	start := len(prefix)
	end := len(path) - len(suffix)
	if start >= end {
		return ""
	}
	return path[start:end]
} 